pilot/pkg/xds/endpoint_builder.go

```diff
@@ -370,3 +386,139 @@ func buildEnvoyLbEndpoint(e *model.IstioEndpoint) *endpoint.LbEndpoint {
 
 	return ep
 }
+
+// TODO this logic is probably done elsewhere in XDS, possible code-reuse
+type mtlsChecker struct {
+	push *model.PushContext
+
+	// cache of host identifiers that have mTLS disabled
+	mtlsDisabledHosts map[string]bool
+
+	// cache of labels/port that have mTLS disabled by peerAuthn
+	peerAuthDisabledMTLS map[string]bool
+
+	svcPort         int
+	destinationRule *networkingapi.DestinationRule
+	// true if the default traffic policy disables mTLS
+	subsetPolicyDisabledMTLS map[string]bool
+	// cache of labels that have mTLS disabled by a subset traffic policy
+	disaledByDestinationRule bool
+}
+
+func newMtlsChecker(push *model.PushContext, svcPort int, dr *config.Config) *mtlsChecker {
+	var drSpec *networkingapi.DestinationRule
+	if dr != nil {
+		drSpec = dr.Spec.(*networkingapi.DestinationRule)
+	}
+	return &mtlsChecker{
+		push:                     push,
+		svcPort:                  svcPort,
+		destinationRule:          drSpec,
+		mtlsDisabledHosts:        map[string]bool{},
+		peerAuthDisabledMTLS:     map[string]bool{},
+		subsetPolicyDisabledMTLS: map[string]bool{},
+		disaledByDestinationRule: mtlsDisabledDefaultTrafficPolicy(dr, svcPort),
+	}
+}
+
+// mTLSDisabled returns true if the given lbEp has mTLS disabled due to any of:
+// - disabled tlsMode
+// - DestinationRule disabling mTLS on the entire host or the port TODO handle subsets
+// - PeerAuthentication disabling mTLS at any applicable level (mesh, ns, workload, port)
+func (c *mtlsChecker) isMtlsDisabled(lbEp *endpoint.LbEndpoint) bool {
+	if disabled, ok := c.mtlsDisabledHosts[lbEpKey(lbEp)]; ok {
+		return disabled
+	}
+	return false
+}
+
+// checkEndpoint checks destination rule, peer authentication and metadata to determine if mTLS was turned off.
+// This must be done during conversion from IstioEndpoint since we still have workload metadata.
+func (c *mtlsChecker) checkEndpoint(ep *model.IstioEndpoint) {
```

Comment:
nit: seems like this should be called something that indicates we are actually persisting this info, not just checking it. I was a bit surprised it was not a `bool` return function